1.9 / 1.9.2 / 1.9.3 / 1.9.4 protocol support

Pokechu22 opened the pull request:
This updates to the 1.9 protocol, and also supports 1.9.1 and 1.9.2 (which have separate protocol IDs).  This update includes the new chunk format.  It does _not_ add any new blocks or items.  It also makes a few other changes:

* Fixed a bug with CheckBasicStyle.lua that caused it to not properly strip line endings, causing issues when used in Cygwin
* Add a `/blk` command for debugging (which was used to figure out what was supposed to be what when trying to diagnose highly broken chunks)
* Split entity mounting/dismounting into separate functions, because 1.9's set passengers packet works with the ridden entity rather than the rider and thus wouldn't be portable with the old method (where the ridden entity would be null when dismounting).

There's a few things that I know are not quite finished yet:

* Spawn eggs don't work since they have different IDs now
* The offhand slot is broken and doesn't work at all.  I haven't tried to write code to deal with it since that's a more complex part of the project that I don't quite understand
* The "[Effect](http://wiki.vg/Protocol#Effect)" packet (the one that does particles and sounds) uses new IDs in 1.9; I haven't tried to update the IDs used in the project since they are in a single enum which doesn't have a good way of converting.
* UUIDs were added to a bunch of packets, but since there isn't a full UUID generated, I use a very non-unique UUID generated from the entity's ID.  The client doesn't seem to care, but this could cause issues.
* I suspect that the new slot in the brewing inventory will cause things to not work properly in the brewing stand (if the brewing stand is implemented), but have not tested it.
* Boats don't seem to work
* No palette is created for chunks, meaning that all blocks are sent with the full 13 bits.  This is more of a performance issue than anything, and since the whole chunk section is sent even if the upper parts are empty, may dramatically reduce performance in comparison to what could be possible.

The main technique I used was following the [Protocol History](http://wiki.vg/Protocol_History) page, rather than trying to rewrite the entire protocol from scratch.  Of course, I wrote the protocol history page a bit ago by using [Burger](https://github.com/mcdevs/Burger)... which I had updated even before that.  So I used a lot of information that wasn't there a month or so ago.  By using protocol history, I could work version-by-version and test along the way, rather than making all the changes and hoping it works.

For the chunk format info, wiki.vg's article is wrong.  I'm going to rewrite sometime later, but the method I used here is based off of what is used by the game itself.

@LogicParrot edit:
Below are some testing binaries.
- [Linux x64](https://builds.cuberite.org/job/Cuberite%20Linux%20x64%20ingame-testing-needed/17/artifact/Cuberite.tar.gz)
- [Windows x64](https://builds.cuberite.org/job/Cuberite%20Windows%20x64%20ingame-testing-needed/16/artifact/Install/Cuberite.zip)
- [Raspberry Pi](https://builds.cuberite.org/job/Cuberite%20Linux%20raspi-armhf%20ingame-testing-needed/16/artifact/Cuberite.tar.gz)

bearbin commented:
The build failed :( 

https://travis-ci.org/cuberite/cuberite/jobs/121871157#L932

sgtbigman commented:
Excellent work! It seems you and I should have been in contact more since I was struggling with the new chunk data serialization. :)

As to why boats are probably not working, there is a new [Steer Boat](http://wiki.vg/Protocol#Steer_Boat) packet, separate from the usual Steer Vehicle one.

If it can be of any help (ignoring my chunk data code, which is a few days behind my further attempts anyways), feel free to check out the [19-protocol-support branch](https://github.com/cuberite/cuberite/compare/master...sgtbigman:19-protocol-support) on my fork. There may be nothing new/helpful on there, but I figured I'd make you aware of it anyways.

Again, excellent work! Also, thanks for your work on wiki.vg, it made working on the new protocol a lot easier, even if my efforts only ended up being a personal learning process.

Pokechu22 commented:
Build errors fixed, hopefully.  Those ones were not generated in visual studio.

Per boats, something weirder is happening - the boat just does not place when used.  Unless I'm doing something wrong with 1.9 boats.

Pokechu22 commented:
OK, I think I _actually_ fixed it.  I did not know that one needed to edit `CMakeLists.txt` if they added a file (I thought that `CMakeLists.txt` was an automatically-generated file and didn't look in it).

Pokechu22 commented:
OK, it's failing now due to an undocumented API constant, `DIG_STATUS_SWAP_ITEM_IN_HAND`.  I don't know how to document an API constant, and from looking at [the list of undocumented constants](http://apidocs.cuberite.org/_undocumented.lua) (found that URL in `main_APIDump.lua`), the other `DIG_STATUS` constants are also undocumented.  What should I do here?

Also, I'll note that VarInts are signed ints, not unsigned.  (`WriteVarInt32` treats them as unsigned, which might cause problems and is generating warnings here)

bearbin commented:
It seems you have the build fixed, you just need to document your new functions since they are exported to the lua API: https://travis-ci.org/cuberite/cuberite/jobs/121931520#L2052

bearbin commented:
You need to add the functions to Server/Plugins/APIDump/APIDesc.lua

You need to add them here: https://github.com/cuberite/cuberite/blob/master/Server/Plugins/APIDump/APIDesc.lua#L2966

You could add the other undocumented ones while you're at it too :)

Pokechu22 commented:
> [17:37:50] Can't load plugin APIDump because of a load error in file Plugins/APIDump/APIDesc.lua: 3 (Plugins/APIDump/APIDesc.lua:2973: unfinished string near '" },')

Well, I suppose that's one way to keep APIDump from causing problems.  Shouldn't be hard to correct; also `BLOCK_FACE` constants shouldn't link.

Pokechu22 commented:
Build is successful now and docs have been fixed.  Anything else I should do?

Pokechu22 commented:
If it'll help with reviewing, [here's a diff between Protocol18x.cpp and Protocol19x.cpp](https://gist.github.com/Pokechu22/b08d71d42bda3e481cf9b62d2f9917b5).

LogicParrot commented:
>the other DIG_STATUS constants are also undocumented. What should I do here?

I think that previously documentation was not strictly enforced, but now it is, and everything *new* must be documented.

>Build is successful now and docs have been fixed. Anything else I should do?

Yay :)

@madmaxoft should probably review this before merge, but rumours say his soul has been devoured by Factorio, and this might take a while.

Also, you should rebase your branch in order to bring in the latest changes from master before this can be merged. This is automatically enforced and we cannot merge before it's done.

LogicParrot commented:
I'll create some binaries, in case someone wants to test this.

Pokechu22 commented:
Rebased and force-pushed, though I don't think it will be a good idea to continuously rebase after reviewing actually starts (it'll make commit comments confusing); until reviewing is done I'll probably wait before rebasing a second time (but will rebase after that).

LogicParrot commented:
Agreed.

mathwhiz1212 commented:
If someone could point me to test code that I could compile, I can help test it.

bearbin commented:
@mathwhiz1212 

There are links to binaries in the first post.

mathwhiz1212 commented:
Can I just pull this and compile it?
https://github.com/Pokechu22/cuberite/tree/protocol-1.9

bearbin commented:
Yes, that's possible.

mathwhiz1212 commented:
@bearbin Ok. Thanks!

mathias-github commented:
Items with right-click features only work when a block is focused.

Seadragon91 commented:
@mathias-github your last comment could be this issue #3130.

Pokechu22 commented:
> Items with right-click features only work when a block is focused.

I haven't implemented [packet 0x1D](http://wiki.vg/Protocol#Use_Item) yet, which probably causes that.

> Command blocks lack animations and sounds when placing and breaking them. The command block GUI doesn't open.

They use a different plugin channel, which I haven't implemented yet.  It should still be possible to implement it, though.  I'm not sure why the GUI is not appearing - my guess is that the ID for command blocks changed and it's not actually being placed correctly, but I'm 99% sure that the GUI is client-side and the server doesn't open it.

> Bats are not animated

No idea why that's happening.  My guess is that there's more entity metadata that isn't known yet.

> Potions turn into water bottles

I think that they made some large changes to the potion ID format in 1.9 (just like with mob eggs).  I'll look into this, but it might be hard to convert the potion into the proper format.

> Bows and fishing rods are broken

Will check this, but I'm guessing it's caused by the same thing as the first item on your list (packet 1D not being implemented).  Do they work while looking at the ground?

mathias-github commented:
> Will check this, but I'm guessing it's caused by the same thing as the first item on your list (packet 1D not being implemented). Do they work while looking at the ground?

Bows work when looking at the ground, but fishing rods don't.

Here's a screenshot of the bats:
![2016-04-11_18 22 48](https://cloud.githubusercontent.com/assets/8754153/14432367/877a2688-0012-11e6-87e2-53a9c454dbe3.png)

Pokechu22 commented:
For bats, I see what's going on, and it's just a typo:

```diff
-			a_Pkt.WriteBEUInt8(0x10);
-			a_Pkt.WriteBEUInt8(Bat.IsHanging() ? 1 : 0);
+			a_Pkt.WriteBEUInt8(11);  // Index 11: Bat flags - currently only hanging
+			a_Pkt.WriteBEUInt8(METADATA_TYPE_BYTE);
+			a_Pkt.WriteBEInt8(Bat.IsHanging() ? 0 : 1);
```

`a_Pkt.WriteBEInt8(Bat.IsHanging() ? 0 : 1);` should be `a_Pkt.WriteBEInt8(Bat.IsHanging() ? 1 : 0);`.  A quick look doesn't show this mistake anywhere else.

With the item frames and other projectiles, what kind of message is seen with the disconnect?

mathias-github commented:
> With the item frames and other projectiles, what kind of message is seen with the disconnect?
Here's the message:
![screenshot from 2016-04-11 19 14 01](https://cloud.githubusercontent.com/assets/8754153/14434123/df2211c8-0019-11e6-8541-0921fde5bc10.png)

Pokechu22 commented:
I see what's happening there too:

```c++
void cProtocol190::SendSpawnObject(const cEntity & a_Entity, char a_ObjectType, int a_ObjectData, Byte a_Yaw, Byte a_Pitch)
{
	ASSERT(m_State == 3);  // In game mode?
	double PosX = a_Entity.GetPosX();
	double PosZ = a_Entity.GetPosZ();
	double Yaw = a_Entity.GetYaw();
	if (a_ObjectType == 71)
	{
		FixItemFramePositions(a_ObjectData, PosX, PosZ, Yaw);
	}

	cPacketizer Pkt(*this, 0x00);  // Spawn Object packet
	// TODO: Bad way to write a UUID, and it's not a true UUID, but this is functional for now.
	Pkt.WriteBEUInt64(0);
	Pkt.WriteBEUInt64(a_Entity.GetUniqueID());
	Pkt.WriteVarInt32(a_Entity.GetUniqueID());
	Pkt.WriteBEUInt8(static_cast<UInt8>(a_ObjectType));
	Pkt.WriteBEDouble(PosX);
	Pkt.WriteBEDouble(a_Entity.GetPosY());
	Pkt.WriteBEDouble(PosZ);
	Pkt.WriteByteAngle(a_Entity.GetPitch());
	Pkt.WriteByteAngle(Yaw);
	Pkt.WriteBEInt32(a_ObjectData);
	Pkt.WriteBEInt16(static_cast<Int16>(a_Entity.GetSpeedX() * 400));
	Pkt.WriteBEInt16(static_cast<Int16>(a_Entity.GetSpeedY() * 400));
	Pkt.WriteBEInt16(static_cast<Int16>(a_Entity.GetSpeedZ() * 400));
}
```

should be

```c++
void cProtocol190::SendSpawnObject(const cEntity & a_Entity, char a_ObjectType, int a_ObjectData, Byte a_Yaw, Byte a_Pitch)
{
	ASSERT(m_State == 3);  // In game mode?
	double PosX = a_Entity.GetPosX();
	double PosZ = a_Entity.GetPosZ();
	double Yaw = a_Entity.GetYaw();
	if (a_ObjectType == 71)
	{
		FixItemFramePositions(a_ObjectData, PosX, PosZ, Yaw);
	}

	cPacketizer Pkt(*this, 0x00);  // Spawn Object packet
	Pkt.WriteVarInt32(a_Entity.GetUniqueID());
	// TODO: Bad way to write a UUID, and it's not a true UUID, but this is functional for now.
	Pkt.WriteBEUInt64(0);
	Pkt.WriteBEUInt64(a_Entity.GetUniqueID());
	Pkt.WriteBEUInt8(static_cast<UInt8>(a_ObjectType));
	Pkt.WriteBEDouble(PosX);
	Pkt.WriteBEDouble(a_Entity.GetPosY());
	Pkt.WriteBEDouble(PosZ);
	Pkt.WriteByteAngle(a_Entity.GetPitch());
	Pkt.WriteByteAngle(Yaw);
	Pkt.WriteBEInt32(a_ObjectData);
	Pkt.WriteBEInt16(static_cast<Int16>(a_Entity.GetSpeedX() * 400));
	Pkt.WriteBEInt16(static_cast<Int16>(a_Entity.GetSpeedY() * 400));
	Pkt.WriteBEInt16(static_cast<Int16>(a_Entity.GetSpeedZ() * 400));
}
```

It was writing the UUID before writing the EID; due to it being a VarInt this could cause the wrong number of bytes to be read.  That should also be a simple enough fix.

mathias-github commented:
Thanks!

Any idea about the buckets?

> - Water and lava buckets only place client-side blocks
- Can't pick up water and lava with empty bucket

Pokechu22 commented:
Buckets were also caused by the unimplemented Use Item packet.

I've fixed that (and bows, fishing rods, and other things that work like that), and I've also fixed TNT and other objects crashing the client and the wrong animation being used for bats.

mathias-github commented:
Everything is solved now, except for the command block issue. Thanks!

Pokechu22 commented:
FYI, I also encountered the frozen glitch (while testing my fix for the other two issues).  I had debug enabled at that time.  I'm guessing that it's not my fault and instead is #3130.

LogicParrot commented:
The frozen glitch is now fixed. You can have a smoother testing experience if you `git cherry-pick 9bec04ab1e357c7ccbf1208153cd1b0a0c53be1d` or if you rebase.

mathias-github commented:
Boss health bars seem to be missing.

alwaysontop617 commented:
Finally Someone added 1.9
:)

alwaysontop617 commented:
We should start our own game. We can have it written in c++. It can have servers, just like minecraft. maybe even better graphics. Feel free to email me your ideas at admin@myflare.cu.cc. This is a really good idea. We can even make it so cuberite users can connect just fine to these type of servers. We also need to come up for a name of the game, I was thinking of longerSEE. Ohh well, make sure to add me on skype my username is dodiaraculus17. Good Luck :)

Pokechu22 commented:
@alwaysontop617 FYI, there are a couple of [alternative client implementations](http://wiki.vg/Client_List) for Minecraft, one of which is in C and another in Rust.  Both of these can connect to normal (and cuberite) servers.

alwaysontop617 commented:
Well I will start the project.

bearbin commented:
@Pokechu22 what you did on bountysource was right, once this PR is merged you will get the bounty.

alwaysontop617 commented:
Here is my bountysource:

https://salt.bountysource.com/checkout/amount?team=cubersphere

Pokechu22 commented:
> Boss health bars seem to be missing.

Yea, those are going to also be difficult to implement.  Boss bars use [a new packet](http://wiki.vg/Protocol#Boss_Bar) in 1.9, which allows multiple colors and multiple bars at once.  However, that packet might be more difficult to implement so I haven't implemented it yet (it requires a UUID and a few other such things which means that it's not trivial to implement).  So right now I just don't have any health bars sent.

I'm specifically not sure where or how to start sending them, and how to deal with it in 1.8.

alwaysontop617 commented:
@Pokechu22  Can U join my organization. 

https://github.com/cubersphere

mathias-github commented:
One does not simply make a game overnight. :P

SphinxC0re commented:
@alwaysontop617 could you please stop abusing this PR for advertising your game idea. If you want to advertise it, use the forum

LogicParrot commented:
I have cleaned the advertisement. @alwaysontop617 please use [this forum](https://forum.cuberite.org/forum-10.html) for this. Thanks :)

alwaysontop617 commented:
Sorry :

On Thu, Apr 14, 2016 at 10:41 PM, LogicParrot <notifications@github.com>
wrote:

> I have cleaned the advertisement. @alwaysontop617
> <https://github.com/alwaysontop617> please use this forum
> <https://forum.cuberite.org/forum-10.html> for this. Thanks :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cuberite/cuberite/pull/3135#issuecomment-210271390>
>



-- 
Sincerely,
Thomas Wilbur

mathias-github commented:
Gameplay-wise this PR is stable enough now in my opinion, and I haven't noticed any major performance issues so far.

bearbin commented:
@madmaxoft or @LogicParrot merge?

Pokechu22 commented:
I've just rebased, by the way.

LogicParrot commented:
@bearbin I did not review the code at all yet. I will do this as soon as I have time.

I have done some short ingame testing though, and I found some issues. I think only the major ones are merge-blocking. I wil constantly edit this and add more as soon as I find them. I will also review the code, hopefully soon.

Major:
- [ ] Clicking F in debug mode is an assertion failure, this should be handled gracefully.
- [ ] Typing `/portal nether` leads to a completely corrupt world. Below are some shots.

Minor:
 - [ ] Falling creates nether particles after hitting the floor.

Nether screenshots:
![2016-04-16_11 04 01](https://cloud.githubusercontent.com/assets/5030488/14579900/cad2a914-03c3-11e6-9211-019f247b9908.png)
![2016-04-16_11 04 29](https://cloud.githubusercontent.com/assets/5030488/14579901/cfc7f65e-03c3-11e6-9cdf-5688829bd660.png)

![2016-04-16_11 03 17](https://cloud.githubusercontent.com/assets/5030488/14579898/c5a1fd8c-03c3-11e6-87fd-6540a14cf7f8.png)

Seadragon91 commented:
I looked into the brewing.
1) The potions now have all the data value 0, that means no current brewing recipes will working in 1.9. They have now a nbt tag.
3) The brewing stand has a new slot, a fuel slot for 20 brewing operations.



I think we should make a own release for 1.9 and no backward compatibility to 1.7/1.8, to many game breaking changes.

Seadragon91 commented:
@Pokechu22 The spawn eggs have a nbt tag, too. That contains the type of the spawn egg.

Pokechu22 commented:
The nether one is strange, but it seems to be caused by the sending of skylight data in the nether.  I'm not sure of the best way to differentiate dimensions (skylight data should only be sent in the overworld and not the nether/end, but dimension data isn't found in `cChunkDataSerializer`).  This happens when any portal is used.

For the wolf one...  I've got no idea what would do that.  Definitely a severe bug.

For the offhand slot one, I've added a case in the switch statement for that.

For the other wolf issue, I'm not sure what's happening there (but I can see it happening).  It looks like the collar isn't rendering at all...  Also it looks like the ownership state isn't updating, but other parts of the wolf state _are_ updating (angry, at least), so that doesn't quite make sense.

For the particles, that's due to effect IDs being different and is still hard to fix.  I'm aware of it but don't have a good solution.  (The particles aren't bad, but the associated explosion noise kinda is.  It's using the "destroy ender crystal" effect).

LogicParrot commented:
Did anyone manage to reproduce the wolf one? It seems rather pecular and could specific to some environment (gpu, drivers, etc.)

Pokechu22 commented:
I did reproduce the wolf one.  No clue why that's happening.

Pokechu22 commented:
I've added support for spawn eggs and potions with the 1.9 NBT format.  (I still haven't implemented the new brewing stand slot, though)

LogicParrot commented:
>I did reproduce the wolf one. No clue why that's happening.

Perhaps the server is sending some malformed death particle or sound effect? (Wild guess)

Pokechu22 commented:
I discovered that you can take screenshots while the screen is black, and they "develop" successfully.

![2016-04-18_15 18 36](https://cloud.githubusercontent.com/assets/8334194/14621656/441372a2-0579-11e6-9fd1-b5cf956f70b6.png)

Where "successfully" is defined as whether looking like it's from a creepypasta is normal.

The wolf also doesn't have to actively be attacking it for this to happen, only targeting.

Pokechu22 commented:
In the client log:

>     [15:39:36] [Client thread/ERROR]: Couldn't render entity
>     java.lang.ClassCastException

Not too descriptive, but I think I see what's happening.

Pokechu22 commented:
It happens whenever the wolf takes damage.  Also this time it decided to give me a useful error message:


>     [16:01:09] [Client thread/ERROR]: Couldn't render entity
>     java.lang.ClassCastException: java.lang.Byte cannot be cast to java.lang.Integer
>         at wj.dk(SourceFile:446) ~[1.9.2.jar:?]
>         at bue.a(SourceFile:26) ~[1.9.2.jar:?]
>         at bue.a(SourceFile:10) ~[1.9.2.jar:?]
>         at bsa.a(SourceFile:401) ~[1.9.2.jar:?]
>         at bsa.a(SourceFile:174) [1.9.2.jar:?]
>         at bsd.a(SourceFile:41) [1.9.2.jar:?]
>         at bte.a(SourceFile:33) [1.9.2.jar:?]
>         at bte.a(SourceFile:9) [1.9.2.jar:?]
>         at brj.a(SourceFile:338) [1.9.2.jar:?]
>         at brj.a(SourceFile:325) [1.9.2.jar:?]
>         at bnl.a(SourceFile:601) [1.9.2.jar:?]
>         at bnd.a(SourceFile:1260) [1.9.2.jar:?]
>         at bnd.b(SourceFile:1165) [1.9.2.jar:?]
>         at bnd.a(SourceFile:1012) [1.9.2.jar:?]
>         at bcc.av(SourceFile:985) [1.9.2.jar:?]
>         at bcc.a(SourceFile:392) [1.9.2.jar:?]
>         at net.minecraft.client.main.Main.main(SourceFile:124) [1.9.2.jar:?]

I think I understand what's happening now.  The wolf's colar render layer is getting that specific exception (caused by an invalid metadata type, by the way - fixing that), and thus the GL state doesn't get reset correctly.  Specifically, the color isn't reset, so if the wolf is taking damage, _everything_ gets the damage hue.  For whatever reason that becomes black on-screen but red in screenshots.

Pokechu22 commented:
Ok, wolf collars (and the wolf graphical glitches) have been fixed.  Right-clicking on a wolf currently isn't working for me, though, but it seems to be a case of 2 right-click packets getting sent at once (causing the wolf to sit and then unsit immediately).

Pokechu22 commented:
And the nether terrain should be fixed now too.  (Also, the end _should_ be fixed, but end generation seems weird so I can't completely verify).

LogicParrot commented:
Great work :)

SphinxC0re commented:
An error is occurring whilst compiling with clang:
![](http://storage8.static.itmages.com/i/16/0419/h_1461053976_1270428_c686405eb5.png)

Pokechu22 commented:
Ugh, I thought I fixed that one.

mathias-github commented:
Anyone still experiencing problems, or is this good gameplay-wise?

LogicParrot commented:
Testing binaries updated. (See first post to download them).

tonibm19 commented:
I get crash in flame.ga server whem moving to a specific zone.
Maybe is optifine. Who knows

Pokechu22 commented:
@tonibm19 Can you post the crash report?  I know how to interpret them.

sgtbigman commented:
Unfinished, but I started some work on a [branch](https://github.com/Pokechu22/cuberite/compare/protocol-1.9...sgtbigman:add-steer-boat-and-vehicle-move-support) towards implementing the Steer Boat and Vehicle Move packets.

Any other functionality not related to those two packets seems to work correctly :)

Pokechu22 commented:
Nice!  FYI, I did a bit of research on what steer boat does exactly and it seems like it mainly is used to [change the boat metadata fields 9 and 10](http://wiki.vg/Entities#Boat).

tonibm19 commented:
[crash-2016-04-23_11.58.31-client.txt](https://github.com/cuberite/cuberite/files/233089/crash-2016-04-23_11.58.31-client.txt)

sgtbigman commented:
Yes, I think I have that in place in the `WriteMetadata` method, but I don't have it updating anything server-side on specific boat objects, just yet.

The steering 'works', although I have the values hardcoded at the moment.

As for the VehicleMove packet, I must not be updating the correct values, since moving in a boat client-side does not work (it just jumps back to the last position, stutters, etc.).

Pokechu22 commented:
@tonibm19: Looks like someone placed a skull with invalid metadata.  

Doesn't seem like it's my bug; [hopper](http://hopper.minecraft.net/crashes/minecraft/MCX-2762446/) lists several other reports on both vanilla and modded servers.  The crash report does give a block coordinate so I can make use of the fact that I'm OP on that server to use worldedit to remove that specific skull.

I also saved a schematic of it so I can investigate further.

tigerw commented:
> Doesn't seem like it's my bug; hopper lists several other reports on both vanilla and modded servers...

With your knowledge of Vanilla, I daresay you've worked on Minecraft :P

mathias-github commented:
The skull crash happens when someone creates a skull on a server with authentication mode off.


Someone's PR changed how skulls are made. Before, you could create a skull of any player, now the server grabs the skull texture from players connected to the server. Since skins don't work when authentication mode is off, a crash occurs.

tonibm19 commented:
Merge? 
I havent seen any big issues so far

Schwertspize commented:
Does this add 1.9 to the existing 1.7 and 1.8 or does it drop 1.8 (and 1.7) completely?

Pokechu22 commented:
It adds 1.9 without changing 1.8 or 1.7.

Seadragon91 commented:
Brewing still doesn't work.
I see that the water bottle has this value `373:8192:8192`, the data value gets added twice.

Pokechu22 commented:
I think that water bottles are incorrectly read as mundane potions would be (water bottles _should_ be `373:0:0`).  I'll look into that; I think it's caused by the "can become splash potion" bit getting set.

However, the sync error is separate - I'm not handling the extra slot that was added to brewing stands, and likely that's causing confusion.  I'm not entirely sure how inventories are handled but I think it's possible to fix it, somehow.

Pokechu22 commented:
Also, when I rebase, does it seem like a good idea for me to insert a commit at the very start that just copies `Protocol18.cpp` to `Protocol19.cpp` to make diffs clearer for the following commits (rather than having that copying merged into 15w31a commit (and obscuring the changes thereof))?

Pokechu22 commented:
Rebased; also a commit copying Protocol18x to Protocol19x has been inserted for readability of the diffs.

mathias-github commented:
About right-clicking a wolf: When holding an item with a right-click feature (e.g. a spawn egg), you can command the wolf to stand/sit. Otherwise it doesn't work.

mathias-github commented:
Minor bug: Thrown splash potions always look like splash water bottles, and use the particles for water bottles.

Pokechu22 commented:
> Minor bug: Thrown splash potions always look like splash water bottles, and use the wrong particles when hitting the ground.

Will fix.  

> Currently, water bottles and splash water bottles seem to be mixed up. Water bottles throw splash water bottles, while splash water bottles don't do anything.

That's going to be hard to fix because cuberite's potion logic is messed up.  In fact, that happens in 1.8 as well.  It's due to `cEntityEffect::IsPotionDrinkable` being used to determine if potions are splash - if the potion isn't drinkable then it's splash.  But the bit that `IsPotionDrinkable` uses is actually _can become a splash potion_.  `cEntityEffect` needs to be adjusted to use the correct bits (13 for can become splash, 14 for is splash).  Problem is that old potions probably don't have the right bits either.

> I'm not sure if this bug is related to this PR, but the effect of Speed II potions keeps increasing if you drink many of these potions (you go faster and faster). When drinking milk, the speed effect doesn't go away.

That also happens with 1.8.  Easy to reproduce if you spam splash potions instead.  My guess is that the remove effect packet isn't properly being sent when the new effect is added, causing them to stack, but I don't know for sure.

> Jukeboxes don't play music (I guess this is also due to effect ID changes in 1.9?)

Yea, that would also be effect ID changes.

Pokechu22 commented:
Potions were broken because they changed from having their effect on the Spawn Object packet's field to a new entity metadata field for the potion.  I've implemented that.

However, the splash effect from a potion is... you guessed it... an effect.  Its ID actually is the same as in 1.8, but the potion effect IDs it uses are different unfortunately.  So the splash particles will be the wrong color.

There's a few other entities that have unimplemented metadata, but they aren't implemented for the other protocols some I'm not going to do them yet.

mathias-github commented:
Currently getting an error while compiling:
```
[ 70%] Built target Bindings
[ 72%] Built target WorldStorage
[ 78%] Built target Generating
[ 84%] Built target Entities
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp: In member function ‘void cProtocol190::WriteItem(cPacketizer&, const cItem&)’:
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effNoEffect’ not handled in switch [-Wswitch]
    switch (Type)
           ^
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effHaste’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effMiningFatigue’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effNausea’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effResistance’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effBlindness’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effHunger’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effWither’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effHealthBoost’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effAbsorption’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3269:11: warning: enumeration value ‘effSaturation’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp: In member function ‘void cProtocol190::WriteEntityMetadata(cPacketizer&, const cEntity&)’:
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3564:81: error: ‘const class cSplashPotionEntity’ has no member named ‘GetItem’
      WriteItem(a_Pkt, reinterpret_cast<const cSplashPotionEntity &>(Projectile).GetItem());
                                                                                 ^
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp: In member function ‘void cProtocol190::WriteMobMetadata(cPacketizer&, const cMonster&)’:
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtInvalidType’ not handled in switch [-Wswitch]
  switch (a_Mob.GetMobType())
         ^
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtBlaze’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtCaveSpider’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtEnderDragon’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtGiant’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtGuardian’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtIronGolem’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtMooshroom’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtSilverfish’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtSnowGolem’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtSpider’ not handled in switch [-Wswitch]
/home/mathias/cuberite/src/Protocol/Protocol19x.cpp:3622:9: warning: enumeration value ‘mtSquid’ not handled in switch [-Wswitch]
src/Protocol/CMakeFiles/Protocol.dir/build.make:206: recipe for target 'src/Protocol/CMakeFiles/Protocol.dir/Protocol19x.cpp.o' failed
make[2]: *** [src/Protocol/CMakeFiles/Protocol.dir/Protocol19x.cpp.o] Error 1
CMakeFiles/Makefile2:1310: recipe for target 'src/Protocol/CMakeFiles/Protocol.dir/all' failed
make[1]: *** [src/Protocol/CMakeFiles/Protocol.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

mathias-github commented:
Thrown potions still look like water bottles after your latest changes.

Pokechu22 commented:
Odd.  They _were_ working; they stopped working; and now they're working again.  I didn't make any changes between them being broken and working (though I did rebase and make a patch to fix the compilation error, they were still broken after that).

Pokechu22 commented:
Also, I'm currently getting a Read Access Violation when using `/stop`, in `cChunkData::Free`.

    Exception thrown: read access violation.
    this->m_Pool. was 0xDDDDDDDD.

I'm not sure if I caused that or not, but I don't really know how to debug it.

LogicParrot commented:
@Pokechu22 Something similar happened in #3151 which caused this. I wasn't sure I caused this either, and I fixed it. See [the cChunkMap destructor in that PR](https://github.com/cuberite/cuberite/pull/3151/files#diff-0fb1fa1e2b709508fa3c9195a46c648bR55).

LogicParrot commented:
Note that I am not entirely sure the issue is identical.

LogicParrot commented:
Does merging #3176 into your branch resolve this?

Pokechu22 commented:
I did a `git merge --no-commit LogicParrot/stopCrashFix` and yea, the bug seems to be fixed (won't push the fixed version on top of my code; I'll wait for that to be merged and then rebase).

For CI: ugh, I really don't know what I'm doing with the API docs.  Didn't I add the doc for `cSplashPotionEntity::GetItem`?  Is there some other place I need to add it?  (And where are the other `cSplashPotionEntity` methods?)

Pokechu22 commented:
... but now potions aren't working again.

LogicParrot commented:
I merged the fix into master.

mathias-github commented:
Once the API docs are fixed, this PR seems ready to me. Someone should review the code though.

mathias-github commented:
One more thing; could you add support for Minecraft 1.9.3 (version number 110)?

Pokechu22 commented:
I'm still not entirely sure _how_ to fix the API docs.  I'd think that e2f1cd6 would fix it, but maybe I put something in the wrong place?  Is there an obvious mistake I'm making?

> One more thing; could you add support for Minecraft 1.9.3 (protocol version number 110)? It will probably be released quite soon.

Planning on doing that; I'm just waiting to see if there are any significant changes between the prerelease and the actual 1.9.3.  There is one thing that would need to be handled, and that is sending tile entities within the chunk itself (though actually, that step can be skipped and tile entities can still be added as they previously were).

The only changes that _really_ are necessary to support 1.9.3 as-is are changing sign sending to use the Update Block Entity packet and shifting the changed packet IDs.

-----

Thinking about it, though, should I also fix the server list ping issue?  The legacy ping format is still being used which causes cuberite servers to not show up in the server list with the wrong version.  More specifically, the wrong disconnect packet is used.  It would require some tweaks to `cProtocolRecognizer` and possibly require removing legacy ping support (which could later be re-added).

LogicParrot commented:
FreeBSD builds of this PR are failing. [See log](https://builds.cuberite.org/job/Cuberite%20FreeBSD%20x64%20ingame-testing-needed/2/console). I haven't really checked if this is specific to this PR, but the Master builds of yesterday have been a success.

LogicParrot commented:
The binaries were updated to the latest commits.

mathias-github commented:
> Thinking about it, though, should I also fix the server list ping issue? The legacy ping format is still being used which causes cuberite servers to not show up in the server list with the wrong version. More specifically, the wrong disconnect packet is used. It would require some tweaks to cProtocolRecognizer and possibly require removing legacy ping support (which could later be re-added).

If you have the time to make these changes, why not?

LogicParrot commented:
@bearbin suspects the FreeBSD failure might have to do with CI running in debug mode, and not something specific with this branch. I suppose this issue can be ignored for now.

LogicParrot commented:
I suggest you neutralize the fall particles at `cPawn::404`, because currently they are the wrong particles.

Pokechu22 commented:
I fixed the API doc issue (I think).  I was editing the wrong file (I didn't realize that some of the documentation was in different files).  Working on the other parts now (disabling those particles makes sense).

Pokechu22 commented:
For my reference, dab398d5d663332527e57a7e239d223f33f4eb77 is when the first 1.7 protocol stuff was added (and the lengthened protocol related things)

Pokechu22 commented:
... and 8d55cd409e38d8759b61e89a4b3dea678c1e5f98 is when the 1.6 protocol code was fully removed.

mathwhiz1212 commented:
Is this fully functional as for 1.9.0-1.9.2?

Pokechu22 commented:
It's got all of the protocol stuff (aside from boats) implemented, if that's what you mean.  There may be bugs that just haven't been found yet, though.  (Also, things like the changed effect IDs haven't been implemented, so it's not perfect).

mathwhiz1212 commented:
Ok. Thank You.

Pokechu22 commented:
Will add 1.9.3/1.9.4 support later today.

Pokechu22 commented:
1.9.3/1.9.4 support has been added.

SphinxC0re commented:
@LogicParrot could you update the binaries

alwaysontop617 commented:
Funny yoyr website says download from 1.7 to 1.8 it should be from 1.8 to
1.9. :D

On Wed, May 11, 2016 at 6:11 AM, Julian Laubstein <notifications@github.com>
wrote:

> @LogicParrot <https://github.com/LogicParrot> could you update the
> binaries
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/cuberite/cuberite/pull/3135#issuecomment-218429333>
>



-- 
Sincerely,
Thomas Wilbur

SphinxC0re commented:
Thats probably because this PR isn't merged yet :-P

SphinxC0re commented:
Also, I don't think we'll drop 1.7 very soon.

tonibm19 commented:
I'd be nice to add a (beta) 1.9 in the website.

LogicParrot commented:
I updated the binaries and the PR title.

mathias-github commented:
So, did anyone find additional bugs? I think we could merge this PR quite soon, if the code looks alright to others.

SphinxC0re commented:
Nice. @LogicParrot what is the status of the freeBSD builds?

bearbin commented:
@SphinxC0re The FreeBSD build is still failing, but I think that's probably because of a underlying issue compiling in Debug mode on FreeBSD, and not something to do with this PR.

SphinxC0re commented:
Let's merge this. This wont get any better than this

mathias-github commented:
I agree. @SphinxC0re Additional features can be added later.

bearbin commented on Server/Plugins/APIDump/APIDesc.lua:
```diff
@@ -2965,6 +2965,28 @@ end
+				DIG_STATUS_DROP_HELD = { Notes = "The player has dropped a single item using the Drop Item key (default: Q)" },
+				DIG_STATUS_DROP_STACK = { Notes = "The player has dropped a full stack of items using the Drop Item key (default: Q) while holding down a specific modifier key (in windows, control)" },
+				DIG_STATUS_SHOOT_EAT = { Notes = "The player has finished shooting a bow or finished eating" },
+        DIG_STATUS_SWAP_ITEM_IN_HAND = { Notes = "The player has swapped their held item with the item in their offhand slot (1.9)" },
```
The indentation is wrong here.

Pokechu22 replied:
Fixed

bearbin commented:
I agree, other bugs can be worked out.

bearbin commented:
@LogicParrot Merge?

SphinxC0re commented:
This looks pretty good and also well documented to me. I'd say merge

LogicParrot commented:
If you think it's ready for merge, merge it.
I won't click that button on anything I haven't thoroughly reviewed :P
That is not to say I oppose merge, it does look pretty good and well documented.

SphinxC0re commented:
I just wanted to hear some more opinions

LogicParrot commented:
My opinion is that it looks very well written. I just didn't have time to completely review it yet and I won't have it any time soon. I don't mind a merge.

LogicParrot commented:
It's worth noting that there are some regressions: Wolves do not sit, brewing, etc. So we should still advertise 1.8 as the recommended protocol until those issues are fixed.

LogicParrot commented:
@mathias-github It might be wise to do this at the homepage. Right now it is implied that 1.9 if fully working.
